qiime2/sdk/usage.py

```diff
@@ -0,0 +1,349 @@
+                             'output(s): %r' % (extra, ))
+
+    def prepare(self, action_signature, scope):
+        opts = {}
+
+        for output in action_signature.outputs.keys():
+            opts[self.get(output)] = output
+
+        return opts
+
+
+class ScopeRecord:
+    STATE_PENDING = 'pending'
+    STATE_FINISHED = 'finished'
+
+    def __init__(self, id: str, factory: callable = None,
+                 value: object = None,
+                 assert_has_line_matching: callable = None):
+        if factory is not None and not callable(factory):
+            # TODO
+            raise Exception
+
+        if assert_has_line_matching is not None and \
+                not callable(assert_has_line_matching):
+            # TODO
+            raise Exception
+
+        self.id = id
+        if value is None:
+            self.state = self.STATE_PENDING
+        else:
+            self.state = self.STATE_FINISHED
+        self._factory = factory
+        self._result = value
+        self._assert_has_line_matching_ = assert_has_line_matching
+
+    def __repr__(self):
+        if self.state == self.STATE_PENDING:
+            return 'ScopeRecord<id=%s, state=%s>' % (self.id, self.state)
+        else:
+            return 'ScopeRecord<id=%s, result=%r>' % (self.id, self.result)
+
+    @property
+    def result(self):
+        if self.is_pending:
+            raise Exception
+        return self._result
+
+    @property
+    def is_pending(self):
+        return self.state == self.STATE_PENDING
+
+    def resolve(self):
+        if self.is_pending:
+            self._result = self._factory()
+            self.state = self.STATE_FINISHED
+
+    def assert_has_line_matching(self, label, path, expression):
+        return self._assert_has_line_matching_(self.id, label, path,
+                                               expression)
+
+
+class Scope:
+    def __init__(self):
+        self._records = dict()
+
+    def __repr__(self):
+        return '%r' % (self._records, )
+
+    @property
+    def records(self):
+        return types.MappingProxyType(self._records)
+
+    def push_record(self, factory=None, value=None,
+                    assert_has_line_matching=None):
+        id_ = uuid.uuid4()
+        record = ScopeRecord(id=id_, factory=factory, value=value,
+                             assert_has_line_matching=assert_has_line_matching)
+        self._records[id_] = record
+        return record
+
+    def get_record(self, id):
+        try:
+            return self.records[id]
+        except KeyError:
+            raise KeyError('No record with id: "%s" in scope.' % (id,))
+
+
+class Usage(metaclass=abc.ABCMeta):
+    def __init__(self):
+        self._scope = Scope()
+        self._reverse_lookup = {}
+
+    def init_data(self, ref, factory):
+        override = self._factory_override(ref, factory)
+        record = self._push_record(factory=override)
+        self._reverse_lookup[ref] = record.id
```

Comment:
Should this be the job of `Scope`?

Reply:
I'm just going to drop the reverse lookup, it isn't strictly necessary, and might just be extra surface area to keep clean.